src/mpi4py/util/pool.py

```diff
@@ -0,0 +1,328 @@
+        self.executor.shutdown(wait=True)
+        _async_executor_shutdown(self)
+
+    def __enter__(self):
+        """Return pool."""
+        return self
+
+    def __exit__(self, *args):
+        """Close pool."""
+        self.close()
+
+
+class ThreadPool(Pool):
+    """Pool using threads as workers."""
+
+    Executor = _futures.ThreadPoolExecutor
+
+
+class AsyncResult:
+    """Asynchronous result."""
+
+    def __init__(self, future, callback=None, error_callback=None):
+        """Wrap a future object and register callbacks."""
+        self.future = future
+        done_cb = _functools.partial(
+            _async_future_callback,
+            callback=callback,
+            error_callback=error_callback,
+        )
+        self.future.add_done_callback(done_cb)
+
+    def ready(self):
+        """Return whether the call has completed."""
+        return self.future.done()
+
+    def successful(self):
+        """Return whether the call completed without raising an exception.
+
+        If the result is not ready then raise `ValueError`.
+
+        """
+        if not self.ready():
+            raise ValueError(f"{self!r} not ready")
+        try:
+            return self.future.exception() is None
+        except _futures.CancelledError:
+            return False
+
+    def wait(self, timeout=None):
+        """Wait until the result is available or *timeout* seconds pass."""
+        try:
+            self.future.exception(timeout)
+        except _futures.CancelledError:
+            pass
+        except _futures.TimeoutError:
+            pass
+
+    def get(self, timeout=None):
+        """Return the result when it arrives.
+
+        If *timeout* is not `None` and the result does not arrive within
+        *timeout* seconds then raise `TimeoutError`.
+
+        If the remote call raised an exception then that exception will be
+        reraised.
+
+        """
+        try:
+            return self.future.result(timeout)
+        except _futures.CancelledError:
+            raise TimeoutError() from None
+        except _futures.TimeoutError:
+            raise TimeoutError() from None
+
+
+class ApplyResult(AsyncResult):
+    """Result type of `apply_async()`."""
+
+
+class MapResult(AsyncResult):
+    """Result type of `map_async()` and `starmap_async()`."""
+
+
+def _chunksize(pool, iterable, chunksize):
+    if chunksize is None:
+        chunksize = 1
+        num_workers = getattr(pool.executor, "_max_workers", 0)
+        if num_workers > 0:  # pragma: no branch
+            num_tasks = _operator.length_hint(iterable)
+            if num_tasks == 0:
+                iterable = list(iterable)
+                num_tasks = len(iterable)
+            if num_tasks > 0:
+                quot, rem = divmod(num_tasks, num_workers * 4)
+                chunksize = quot + bool(rem)
+    return iterable, chunksize
+
+
+def _async_future_callback(future, callback=None, error_callback=None):
```

Comment:
I am sensing the standalone helpers `_async*()` are here because you don't want any private methods to appear in the `Pool` class (which I am fine with)?

Reply:
In the most part, yes, I have a tendency to avoid private methods. But IMO, the only one that would make sense to add as a private method is `_chunksize()`. All the others are related to nasty implementation details that I prefer to keep as hidden as possible.